docs/proposals/csi.md

```diff
@@ -0,0 +1,349 @@
+The customers often need to store data and run stateful applications at edge.
+For example, in a system which collects video data and then makes analysis,
+one application stores the video data into a shared storage,
+and another application reads the data from the storage to make analysis.
+KubeEdge should allow customers to store data using StorageClass (SC),
+PersistentVolume (PV) and PersistentVolumeClaim (PVC)
+so that the customers can deploy stateful applications at edge.
+
+The Container Storage Interface (CSI) is a specification that resulted from cooperation
+between community members including Kubernetes, Mesos, Cloud Foundry, and Docker.
+The goal of this interface is to establish a standardized mechanism
+to expose arbitrary storage systems to their containerized workloads.
+CSI Spec has already been released v1.1 and lots of CSI Drivers are released by vendors.
+CSI Volume Plugin is supported beginning with kubernetes v1.9 and in GA status since v1.13.
+
+* [CSI Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+* [CSI Drivers](https://kubernetes-csi.github.io/docs/drivers.html)
+* [CSI Volume Plugin in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+We can support Container Storage Interface (CSI) in KubeEdge,
+and give a standardized solution to the customers who want to use persistent storage at edge.
+The customers can choose any kind of CSI Drivers on demands. Since KubeEdge is built upon Kubernetes,
+it is possible to reuse some of existing framework from Kubernetes in KubeEdge.
+
+## Goals
+
+* To support [basic CSI Volume Lifecycle](https://github.com/container-storage-interface/spec/blob/master/spec.md#volume-lifecycle) in KubeEdge.
+   * Create Volume and Delete Volume.
+   * Controller Publish Volume and Controller Unpublish Volume.
+   * Node Stage Volume and Node Unstage Volume.
+   * Node Publish Volume and Node Unpublish Volume.
+* Compatiable with Kubernetes and CSI.
+
+## Non-goals
+
+* To support Raw Block Volume.
+* To support Volume Snapshot and Restore.
+* To support Volume Topology.
+
+## Proposal
+
+### Requirement
+* Kubernetes v1.13+
+* CSI Spec v1.0.0+
+
+### Architecture
+<img src="../images/csi/csi-architecture.png">
+
+The added components in KubeEdge are including:
+* External-Provisioner: list watch the Kubernetes API Resource `PersistentVolumeClaim` (`PVC`)
+   and send messages to edge to invoke actions including `Create Volume` and `Delete Volume` at edge.
+* External-Attacher: list watch the Kubernetes API Resource `VolumeAttachment` and send messages to edge
+   to invoke actions including `Controller Publish Volume` and `Controller Unpublish Volume` at edge.
+* Volume Controller: create the Kubernetes API Resource `VolumeAttachment` based on the messages from edge.
+* CSI Volume Plugin: invoke actions including `Node Stage Volume` and `Node Unstage Volume`,
```

Comment:
Are all these components developed from scratch ? or reused from the csi work ?
If we are reusing stuff , then it's better we have different color coding schemes which can clearly highlight this.

Reply:
Thanks, I will use different color in the graph:)

Reply:
Done